operator/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+                error!(
+                    "Pod [{}] has no labels, this is illegal, deleting it.",
+                    Meta::name(&pod),
+                );
+                self.context.client.delete(&pod).await?;
+                continue;
+            }
+        }
+
+        self.node_information = Some(NodeInformation {
+            master,
+            worker,
+            history_server: Some(history_server),
+        });
+
+        info!(
+            "Status[current/spec] - {} [{}/{}] | {} [{}/{}] | {} [{}/{}]",
+            SparkNodeType::Master.as_str(),
+            self.node_information
+                .as_ref()
+                .unwrap()
+                .get_pod_count(SparkNodeType::Master),
+            self.spec.master.get_instances(),
+            SparkNodeType::Worker.as_str(),
+            self.node_information
+                .as_ref()
+                .unwrap()
+                .get_pod_count(SparkNodeType::Worker),
+            self.spec.worker.get_instances(),
+            SparkNodeType::HistoryServer.as_str(),
+            self.node_information
+                .as_ref()
+                .unwrap()
+                .get_pod_count(SparkNodeType::HistoryServer),
+            match self.spec.history_server.as_ref() {
+                None => 0,
+                Some(hs) => hs.get_instances(),
+            }
+        );
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+
+    /// Reconcile the cluster according to provided spec. Start with master nodes and continue to worker and history-server nodes.
+    /// Create missing pods or delete excess pods to match the spec.
+    pub async fn reconcile_cluster(&self) -> SparkReconcileResult {
+        trace!("Starting {} reconciliation", SparkNodeType::Master.as_str());
+
+        if let Some(node_info) = &self.node_information {
+            self.reconcile_node(&SparkNodeType::Master, &node_info.master)
+                .await?;
+            self.reconcile_node(&SparkNodeType::Worker, &node_info.worker)
+                .await?;
+
+            if let Some(history_server) = &node_info.history_server {
+                self.reconcile_node(&SparkNodeType::HistoryServer, history_server)
+                    .await?;
+            } else {
+                return Ok(ReconcileFunctionAction::Continue);
+            }
+        } else {
+            return Ok(ReconcileFunctionAction::Done);
+        }
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+
+    /// Reconcile a SparkNode (master/worker/history-server)
+    ///
+    /// # Arguments
+    /// * `node_type` - SparkNodeType (master/worker/history-server)
+    /// * `nodes` - Map where the key is a selector hash and list of pods owned by that selector
+    ///
+    async fn reconcile_node(
+        &self,
+        node_type: &SparkNodeType,
+        nodes: &HashMap<String, Vec<Pod>>,
+    ) -> SparkReconcileResult {
+        // TODO: check if nodes terminated / check if nodes up and running
```

Comment:
I'd prefer if you would not check in commented code, leave it in a local branch if you like (and the TODO is fine as well!)